File: mesh3D/Fortran3D/examples/example_concave.f90
PROGRAM example_concave

    USE TypesDef, ONLY: dp
    USE PolyhedronMesh, ONLY: t_polyhedron, MeshReader
    USE OptimalPolyCuba3D_Module, ONLY: OptimalPolyCuba3D

    IMPLICIT NONE

    !***********************************************************************
    !
    !   Esempio:
    !       Cubatura su dominio poliedrale concavo
    !
    !   Descrizione:
    !       Questo esempio dimostra l'utilizzo del metodo
    !       OptimalPolyCuba3D per l'integrazione di una funzione su un
    !       dominio poliedrale concavo, rappresentato da una mesh
    !       superficiale triangolare.
    !
    !***********************************************************************

    !***********************************************************************
    !   Dichiarazione delle variabili
    !***********************************************************************

    INTEGER                         :: ade
    INTEGER                         :: n_vertici, n_facce
    REAL(dp)                        :: Integrale
    INTEGER                         :: i
    REAL(dp), ALLOCATABLE           :: XYZ(:,:)
    REAL(dp), ALLOCATABLE           :: W(:)
    REAL(dp)                        :: elapsedTime
    REAL(dp)                        :: t_start, t_end

    TYPE(t_polyhedron)              :: poly

    CHARACTER(LEN=256)              :: vertici_file
    CHARACTER(LEN=256)              :: tri_file

    !***********************************************************************
    !   Parametri e caricamento mesh
    !***********************************************************************

    ade = 1

    WRITE(*,'(A)')
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)') '                 OPTIMALPOLYCUBA3D'
    WRITE(*,'(A)') '        Cubatura su Dominio Poliedrale concavo'
    WRITE(*,'(A)') '**************************************************************'
    WRITE(*,'(A)')

    WRITE(*,'(A,I0)') 'Ade:                   ', ade

    ! Percorsi relativi alla cartella examples.
    vertici_file = 'concave_vertex.dat'
    tri_file     = 'concave_tri.dat'

    ! Lettura della mesh superficiale triangolare.
    CALL MeshReader(vertici_file, tri_file, poly)

    n_vertici = SIZE(poly%vertici,1)
    n_facce   = SIZE(poly%facce,1)

    WRITE(*,'(A,I0)') 'Numero di vertici:     ', n_vertici
    WRITE(*,'(A,I0)') 'Numero di facce:       ', n_facce

    !***********************************************************************
    !   Definizione funzione integranda
    !***********************************************************************

    WRITE(*,'(A)') 'Funzione integranda:   f(x,y,z) = 1'

    !***********************************************************************
    !   Inizio regola di cubatura
    !***********************************************************************

    WRITE(*,'(A)')
    WRITE(*,'(A)') 'Inizio Cubatura...'

    CALL CPU_TIME(t_start)

    CALL OptimalPolyCuba3D(ade, poly%vertici, poly%facce, XYZ, W)

    CALL CPU_TIME(t_end)

    elapsedTime = t_end - t_start

    Integrale = 0.0_dp
    DO i = 1, SIZE(W)
        Integrale = Integrale + W(i) * f(XYZ(i,1), XYZ(i,2), XYZ(i,3))
    END DO

    !***********************************************************************
    !   Visualizzazione risultati e punti di cubatura
    !***********************************************************************

    WRITE(*,'(A)') 'Fine Cubatura...'

    WRITE(*,'(A)')
    WRITE(*,'(A,F0.15)') 'Integrale numerico:       ', Integrale
    WRITE(*,'(A,ES14.6,A)') 'Tempo di calcolo:       ', elapsedTime, ' s'

    CONTAINS

    !***********************************************************************
    !   Funzione integranda
    !
    !   Modificare esclusivamente questa funzione per cambiare
    !   l'integranda dell'esempio.
    !***********************************************************************

    REAL(dp) FUNCTION f(x,y,z)

        IMPLICIT NONE

        REAL(dp), INTENT(IN) :: x
        REAL(dp), INTENT(IN) :: y
        REAL(dp), INTENT(IN) :: z

        f = 1.0_dp

    END FUNCTION f

END PROGRAM example_concave

File: Mesh3D/Matlab3D/examples/example_concave.m
clear;
clc;
close all;

projectRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(projectRoot);
addpath(fullfile(projectRoot, 'src'));

%**************************************************************************
%
%   Esempio: 
%       Cubatura su dominio concavo
%
%   Descrizione:
%       Questo esempio dimostra l'utilizzo del metodo OPC3D per
%       l'integrazione di una funzione su un dominio poliedrale concavo,
%       rappresentato da una mesh superficiale triangolare. 
%
%**************************************************************************

%**************************************************************************
%  Parametri e caricamento mesh
%**************************************************************************

ade = 1;

fprintf('\n');
fprintf('**************************************************************\n');
fprintf('                          OPC3D\n');
fprintf('        Cubatura su Dominio Poliedrale Concavo\n');
fprintf('**************************************************************\n');
fprintf('\n');

fprintf('Ade:                   %d\n', ade);

vertices = load('concave_vertex.dat');

facets = load('concave_tri.dat');

fprintf('Numero di vertici:     %d\n', size(vertices,1));
fprintf('Numero di facce:       %d\n', size(facets,1));

%**************************************************************************
% Definizione funzione integranda
%**************************************************************************

f = @(x,y,z) ones(size(x));
f_string = 'f(x,y,z) = 1';
fprintf('Funzione integranda:   %s\n', f_string);

%**************************************************************************
% Plot dominio poliedrale
%**************************************************************************

figure('Color','w');

patch('Vertices', vertices, ...
      'Faces', facets, ...
      'FaceColor', [0.7 0.7 0.7], ...
      'EdgeColor', [0.2 0.2 0.2], ...
      'FaceAlpha', 0.65);

hold on;

axis equal;
grid on;
box on;

xlabel('$x$', 'Interpreter','latex');
ylabel('$y$', 'Interpreter','latex');
zlabel('$z$', 'Interpreter','latex');

title('Dominio Poliedrale Concavo','Interpreter','latex');

view(3);

hold off;

%**************************************************************************
% Inizio regola di cubatura
%**************************************************************************

fprintf('\n');
fprintf('Inizio Cubatura...\n');

[XYZ, W] = OPC3D(ade, vertices, facets, "D");
%[XYZ, W] = OPC3D(ade, vertices, facets, "GJ");
elapsedTime = timeit(@() OPC3D(ade, vertices, facets, "D"));

I = W' * f(XYZ(:,1), XYZ(:,2), XYZ(:,3));

%**************************************************************************
% Visualizzazione risultati e punti di cubatura
%**************************************************************************

fprintf('Fine Cubatura...\n');

fprintf('\n');
fprintf('Integrale numerico:    %.15e\n', I);
fprintf('Tempo di calcolo:      %.6e s\n', elapsedTime);
